### PR TITLE
feat: Make FC reporting in errors nicer

### PR DESCRIPTION
**Pull request by croyzor**

### CODE REVIEW

**Comment by acl-cqc** [comment]
My suggestion: make `show FC` check whether start and end line numbers are the same; if so then display `6:22-31` else `6:22 to 8:31` (or something like that)

**Comment by croyzor** [comment]
> My suggestion: make `show FC` check whether start and end line numbers are the same; if so then display `6:22-31` else `6:22 to 8:31` (or something like that)

Not quite what you're suggesting, but see #39 !

**Comment by acl-cqc** [comment]
> > My suggestion: make `show FC` check whether start and end line numbers are the same; if so then display `6:22-31` else `6:22 to 8:31` (or something like that)
> 
> Not quite what you're suggesting, but see #39 !

That appears to be this PR. Here AFAICS you are showing `start FC` only....?

**Comment by croyzor** [comment]
> My suggestion: make `show FC` check whether start and end line numbers are the same; if so then display `6:22-31` else `6:22 to 8:31` (or something like that)

See #43 for this version. I'll say my preference is for the one in this PR. While we have the information, I feel that reporting the 'start FC' tells the users where to go in the file to find the problem, and to see where it ends they have the code span that is printed by the error message. I'm not convinced printing the 'end FC' adds anything

**Comment by acl-cqc** [comment]
> > My suggestion: make `show FC` check whether start and end line numbers are the same; if so then display `6:22-31` else `6:22 to 8:31` (or something like that)
> 
> See #43 for this version. I'll say my preference is for the one in this PR. While we have the information, I feel that reporting the 'start FC' tells the users where to go in the file to find the problem, and to see where it ends they have the code span that is printed by the error message. I'm not convinced printing the 'end FC' adds anything

Ok, interesting to compare, my thoughts:
* Start of FC is not necessarily where the user needs to look; the problem could be anywhere within the range. (Or indeed outside of it, but we can't really help with that...much/easily).
* However, really (comparing the output in the two PRs) the useful bit in the text message is the line number; the column number is less useful (maybe if you have really, really, long lines, or are otherwise relying on column numbers shown in your text editor to find the offending snippet?). The location within the line is better shown by the highlight-section, as you suggest.
* So, how about `error in function foo on line 6`, `error in function bar on lines 6-8` followed by the highlighted section? (Or, perhaps show the character numbers only if a single line, or something like that, if we want to cover that case)
* Also, super-nit, but I like the line before the code extract to end with a colon `:`

File: brat/Brat/Error.hs [hunk 1]
```diff
@@ -181,7 +181,7 @@ instance Show SrcErr where
   show (SrcErr prelim Err{msg=msg}) = unlines [prelim, "  " ++ show msg]
 
 errHeader :: String -> String
-errHeader name = "Error in " ++ name ++ ":"
+errHeader name = "Error in " ++ name
```
Comment (acl-cqc) [inline brat/Brat/Error.hs:184]: Reinstate the `:` here, it's disappeared from messages that don't have a line number

File: brat/Brat/Error.hs [hunk 2]
```diff
@@ -191,9 +191,17 @@ addSrcContext _ _ (Right r) = Right r
 addSrcContext fname cts (Left err@Err{fc=fc}) = Left (SrcErr msg err)
  where
   msg = case fc of
-    Just fc -> unlines (errHeader (fname ++ '@':show fc):showFileContext cts fc)
+    Just fc -> unlines (errHeader (fname ++ prettyFC fc)
+                       :showFileContext cts fc
+                       )
     Nothing -> errHeader fname
```
Comment (acl-cqc) [inline brat/Brat/Error.hs:197]: (This is one place that's lost a trailing `:` but not the only one, it's also gone from `addSrcName`, hence the suggestion to reinstate it in `errHeader`)

File: brat/Brat/Error.hs [hunk 3]
```diff
@@ -191,9 +191,17 @@ addSrcContext _ _ (Right r) = Right r
...
 
+  prettyFC fc = let Pos startLine _ = start fc
+                    Pos endLine _ = end fc
+                in  if startLine == endLine
+                    then " on line " ++ show startLine ++ ":"
```
Comment (acl-cqc) [inline brat/Brat/Error.hs:202]: ....and then you can drop the `:` from both of these

**Review by acl-cqc — APPROVED** [review]
Nice! :)
Fix the colons and looks good to go, thanks!

**Pull request merged**